### PR TITLE
Allow partial update of downloads in PUT filterOutput

### PR DESCRIPTION
**Pull request by ChrysmOre**
### What

- Updated the Mongo query for updating a filterOutput to allow partial updating of the downloads subdocument.
Before any update of a single piece would replace the entire document, but as each service needs to update a separate aspect of it this was unsuitable.
- Updated old component test now placeholder step has been implemented
### How to review

Check changes make sense, tests pass

### Who can review

Anyone

### CODE REVIEW

File: api/contract.go [hunk 1]
```diff
@@ -66,27 +66,36 @@ type putFilterResponse struct {
 	PopulationType string        `json:"population_type"`
 }
 
-// updateFilterOutputRequest is the request body for POST /filters
-type updateFilterOutputRequest struct {
-	model.FilterOutput
+// putFilterOutputRequest is the request body for POST /filters
+type putFilterOutputRequest struct {
```
Comment (jayfreestone) [inline api/contract.go:70]: Probably can update the comment to say `PUT`.

File: datastore/mongodb/filter_outputs.go [hunk 2]
```diff
@@ -50,30 +50,52 @@ func (c *Client) CreateFilterOutput(ctx context.Context, f *model.FilterOutput)
 func (c *Client) UpdateFilterOutput(ctx context.Context, f *model.FilterOutput) error {
 	col := c.collections.filterOutputs
```
Comment (jayfreestone) [inline datastore/mongodb/filter_outputs.go:51]: Realise you're just editing an existing method, but wonder if it's worth locking the table here.

File: datastore/mongodb/filter_outputs.go [hunk 3]
```diff
@@ -50,30 +50,52 @@ func (c *Client) CreateFilterOutput(ctx context.Context, f *model.FilterOutput)
...
+	queryFilter := bson.M{"id": f.ID}
+
+	if err := c.conn.Collection(col.name).FindOne(ctx, queryFilter, &existing); err != nil {
+		err := &er{
+			err: errors.Wrap(err, "failed to find filter output"),
+		}
+		if errors.Is(err, mongodb.ErrNoDocumentFound) {
+			err.notFound = true
+		}
+		return err
```
Comment (jayfreestone) [inline datastore/mongodb/filter_outputs.go:63]: Probably no need to multiline this? Not that it matters:

```go
return nil, &er{
	err: err,
	notFound: errors.Is(err, mongodb.ErrNoDocumentFound),
}
```

File: datastore/mongodb/filter_outputs.go [hunk 4]
```diff
@@ -50,30 +50,52 @@ func (c *Client) CreateFilterOutput(ctx context.Context, f *model.FilterOutput)
...
+	if f.Downloads.TXT != nil {
+		fields["downloads.txt"] = f.Downloads.TXT
+	}
+	if f.Downloads.XLS != nil {
+		fields["downloads.xls"] = f.Downloads.XLS
+	}
 
-	var rec *mongodb.CollectionUpdateResult
-	sc := bson.M{"id": f.ID}
+	update := bson.M{"$set": fields}
```
Comment (jayfreestone) [inline datastore/mongodb/filter_outputs.go:88]: Do you also need to update the `LastUpdated` field on the output? IIRC it's not a trigger or anything but a field you have to manually set, although I could be misremembering.

File: features/steps/steps.go [hunk 5]
```diff
@@ -265,9 +263,32 @@ func (c *Component) privateEndpointsAreNotEnabled() error {
 	return nil
 }
 
-func (c *Component) theDocumentInTheDatabaseShouldBe(id string, doc *godog.DocString) error {
-	// TODO: implement step for verifying documents stored in Mongo. No prior
-	// art of this being done properly in ONS yet so save to be done in future ticket
+func (c *Component) aDocumentInCollectionWithKeyValueShouldMatch(col, key, val string, doc *godog.DocString) error {
```
Comment (jayfreestone) [inline features/steps/steps.go:266]: 👍 Nice, will borrow for my PR.

File: features/steps/steps.go [hunk 6]
```diff
@@ -265,9 +263,32 @@ func (c *Component) privateEndpointsAreNotEnabled() error {
...
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal bson document")
+	}
+
+	if err := json.Unmarshal(b, &result); err != nil {
+		return errors.Wrap(err, "failed to unmarshal result")
+	}
+
+	if diff := cmp.Diff(result, expected); diff != "" {
+		return fmt.Errorf("-got +expected)\n%s\n", diff)
```
Comment (jayfreestone) [inline features/steps/steps.go:289]: This is a super nitpick, but I think traditionally these are the [other way around](https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff):

```go
if diff := cmp.Diff(want, got); diff != "" {
    t.Errorf("MakeGatewayInfo() mismatch (-want +got):\n%s", diff)
}
```

File: datastore/mongodb/filter_outputs.go [hunk 7]
```diff
@@ -50,30 +50,52 @@ func (c *Client) CreateFilterOutput(ctx context.Context, f *model.FilterOutput)
...
 		return errors.Wrap(err, "failed to update filter output")
 	}
 
-	//This should not happen unless the recored is being removed btween the first find and the update.
-	//Check if the update failed sue to search condition.
-	//Update returns no error but the result's MatchedCount has to be checked for number of updated records
+	// This should not happen unless the recored is being removed bteween the
+	// first find and the update. Check if the update failed sue to search condition.
+	// Update returns no error but the result's MatchedCount has to be checked
+	// for number of updated records
```
Comment (jayfreestone) [inline datastore/mongodb/filter_outputs.go:98]: Nit: few spelling errors here.

File: datastore/mongodb/filter_outputs.go (hunk 4 above)
Comment (ChrysmOre) [reply datastore/mongodb/filter_outputs.go:86]: Added it in the handler

**Review by jayfreestone — APPROVED** [review]

**Review by jayfreestone — APPROVED** [review]

**Pull request merged**